api

```diff
@@ -351,6 +350,94 @@ app_to_pkgname() { #given an app-name, convert it to a unique, valid package-nam
   echo "pi-apps-$(echo "$app" | md5sum | cut -c1-8 | awk '{print $1}')"
 }
 
+add_external_repo() { # add an external apt repo and its gpg key
+  repopath="$1"
+  reponame="$2"
+  pubkeyurl="$3"
+
+  # check if all needed vars are set
+  [ -z "$repopath" ] && error "add_external_packages: repopath not set"
+  [ -z "$reponame" ] && error "add_external_packages: reponame not set"
+  [ -z "$pubkeyurl" ] && error "add_external_packages: pubkeyurl not set"
+
+  # exit if reponame contains space, since apt doesn't accept .list files with spaces in filename or keyname.
+  if [[ $reponame = *" "* ]]; then
+    error "add_external_repo: provided reponame contains a space."
+  fi
+
+  # check if links are valid
+  wget -q --spider "$pubkeyurl" || error "add_external_packages: pubkeyurl isn't a valid link"
+
+  # make pi-apps key storage dir if it doesn't exist
+  if [ ! -d /usr/share/keyrings ]; then
+    sudo mkdir -p /usr/share/keyrings || error "add_external_repo: failed to create pi-apps key storage dir."
+  fi
+
+  # check if .list file already exists
+  [ -f /etc/apt/sources.list.d/${reponame}.list ] && error "add_external_repo: /etc/apt/sources.list.d/${reponame}.list already exists."
+
+  # download gpg key from specified url
+  rm -f /usr/share/keyrings/${reponame}-keyring.gpg
+  curl -fsSL "$pubkeyurl" | sudo gpg --dearmor -o /usr/share/keyrings/${reponame}-keyring.gpg
```

Comment:
Also, `sudo` is used to create the ${reponame}-keyring.gpg file, but no sudo is used to remove it. I haven't tested this, but it seems like rm would fail.